### ISSUE TITLE
Add optional caching?

### ISSUE THREAD

**Issue by TomLisankie** [shogowada/json-rpc-2.0#23]
Over in [rpc-caching-server](https://github.com/Monadical-SAS/rpc-cache-server) we have built (and continue to build) a service that's meant to be used to cache expensive RPCs made to [Solana](https://solana.com/) clusters. We make heavy use of this repository inside of that package.

I was reviewing some of the code of our project today and it occurred to me that a good chunk of our custom code wouldn't have to exist if caching was implemented in some way in this package. Given that we're certainly not the only people who have to make expensive RPCs and need to be able to cache the results, I think it would make sense to implement inside of this package as an option developers can turn on.

What do you think @shogowada? I would be happy to take on this task.

**Comment by TomLisankie** [comment]
Hey, my apologies for not responding sooner, I've had to be offline since the last half of last week. Will give you a proper response sometime this afternoon.

**Comment by TomLisankie** [comment]
Okay so the reason we had gone with Redis instead of just doing it in-memory was because we needed to be able to put multiple cache-readers behind a load balancer. All of these instances needed to be able to access the same cache state. It also helps that most cloud providers have managed caching solutions that use Redis.

So to be honest, I think it would be ideal to have both but think it would be a good first step to implement the external cache usage first since I already know vaguely what I'd need to do and it would be the highest leverage solution for our use-case in the short-term.

What do you think @shogowada?

**Comment by shogowada** [comment]
Thanks for the explanation.

I don't think it makes sense to implement a built-in behavior that's specific to some technology, but I can probably add middleware so that it would be easier to support external cache and share the solution. The middleware will be able to intercept requests so that it can choose to fetch result from cache or do whatever.

So something like this:
```javascript
jsonRPCServer.applyMiddleware(async (next, request) => {
  const cachedResponse = await fetchCache(request);
  return cachedResponse || next(request);
});
```

Or you may choose to do it on client side:
```javascript
jsonRPCClient.applyMiddleware((next, request) => {
  // ...
});
```

The actual signature may not end up looking like this, but would something like this make things easier on your side?

**Comment by TomLisankie** [comment]
Yes! That would be very helpful.

**Comment by shogowada** [comment]
@TomLisankie The server middleware is released as part of v0.2.19. I didn't support client middleware because I couldn't see a good use case yet. The usage is documented here https://github.com/shogowada/json-rpc-2.0#middleware, but let me know if it wasn't clear how to use it.

Thanks for the idea again!
